cpp/src/arrow/filesystem/test_util.cc

```diff
@@ -646,6 +647,114 @@ void GenericFileSystemTest::TestGetTargetStatsSelector(FileSystem* fs) {
   ASSERT_RAISES(IOError, fs->GetTargetStats(s, &stats));
 }
 
+using ::testing::ElementsAre;
+
+FileStats File(std::string path) {
+  FileStats st;
+  st.set_type(FileType::File);
+  st.set_path(path);
+  return st;
+}
+
+FileStats Dir(std::string path) {
+  FileStats st;
+  st.set_type(FileType::Directory);
+  st.set_path(path);
+  return st;
+}
+
+// Utility functions to help testing/debugging
+std::ostream& operator<<(std::ostream& os, const FileStats& stats) {
+  return os << "FileStats(" << stats.type() << ", " << stats.path() << ")";
+}
+
+bool operator==(const FileStats& lhs, const FileStats& rhs) {
```

Comment:
We can probably put this in the public API, but it must compare size and mtime as well.